docs/content/sips/0xx-spin-factors.md

```diff
@@ -0,0 +1,108 @@
+title = "SIP 0XX - Spin Factors"
+template = "main"
+date = "2024-05-20T12:00:00Z"
+
+---
+
+Summary: Refactor Spin Runtime Functionality with "Factors"
+
+Owner: lann.martin@fermyon.com
+
+Created: May 20, 2024
+
+## Background
+
+Spin 1.0 shipped with a mechanism for extending the runtime environment with
+loosely-coupled host functionality called "host components". As Spin has
+evolved, more and more runtime functionality has fallen outside of the scope of
+this mechanism, leading to the code for these features spreading out over the
+Spin codebase. This not only makes it hard to read and understand these features
+but also raises the bar for unfamiliar developers trying to add new features.
+
+Separately, the introduction of the SpinKube project has made it clear that
+there will be (at least) two major embeddings of the Spin runtime environment:
+the Spin CLI (`spin up`) and SpinKube (`containerd-shim-spin`). In order to
+provide better integration into the Kubernetes ecosystem, some runtime feature
+implementations will diverge between these embeddings, making the loose coupling
+of features even more important.
+
+## Proposal
+
+The basic inversion of control approach used by `HostComponent`s will be
+redesigned and expanded into a new system called Spin Factors, where independent
+feature sets are organized into individual "factors". Some of the goals of this
+new system:
+
+- Expand the set integration points that a factor can use to hook into the
+  lifecycle of an application request. These will be driven by feature
+  requirements and may include e.g.:
+  - Runtime startup
+  - Application initialization
+  - Runtime config parsing
+  - Component pre-instantiation
+  - Component instantiation
+  - Cleanup (post-execution)
+
+- Allow loosely-coupled dependencies between "factors" (features). In order to
+  simplify implementation and reasoning about these dependencies the dependency
+  graph must be acyclic.
+
+## Implementation Plan
+
+The overall implementation plan will be to (in parallel):
+
+- Introduce a new `spin-factors` crate containing the basic framework.
+- Refactor existing Spin features (mostly existing `HostComponent`s) to use
+   this new framework.
+   - Framework features will be developed as needed to support this refactoring.
+- Refactor `spin-core`, `spin-trigger`, etc. to use `spin-factors`.
+   - Depending on subjective evaluation at this point, possibly merge
+   `spin-core` and `spin-factors`.
+
+## Implementation Details
+
+Based on initial prototyping, the following Rust types represent the starting
+point for `spin-factors` (note that some type details are elided for clarity):
+
+```rust
+pub trait Factor {
+    // Builder represents a type that can prepare request state for the factor.
+    type Builder: FactorBuilder<Self>;
+    // Data represents request state for the factor.
+    type Data;
```

Comment:
If data === request state, why not call this `RequestState` which while vague is less generic than `Data` which might as well be called `Foo`.

Reply:
This aligns with Wasmtime's internal terminology: https://docs.rs/wasmtime/21.0.0/wasmtime/struct.Store.html#method.data

Reply:
Hmm... by that logic we shouldn't call it a `Factor`, but rather a `Store`. I personally don't think adopting the wasmtime naming convention here helps. I consider myself fairly familiar with the wasmtime API, and I personally find aligning with that naming highly confusing.

Reply:
It's also what [`HostComponent` calls it](https://github.com/fermyon/spin/blob/e078f8d36e43d8ddc7f34c74d8c4d2b37550c9a6/crates/core/src/host_component.rs#L50), though that's not an argument either way in this context.

Happy to bikeshed, though "request state" is also a bad name; I've already changed this comment to reword.

Reply:
`PerInstanceState` would be initial candidate into the bikeshedding arena.

Reply:
I did call it `InstanceState` in one prototype iteration but got sick of typing it over and over :sweat_smile:

Reply:
Maybe even just `State` would be an incremental improvement?

Reply:
`State` is an improvement, but I like the fact that `InstanceState` makes it clear how long the state lives for and what it associated with. `State` could sound like persisted state associated with the app, for example.